basic-loadbalancer-aware-server/src/main/java/com/rtbhouse/grpc/loadbalancer/SingleLoadBalancerConnector.java

```diff
@@ -0,0 +1,335 @@
+        || status == SingleConnectorStatus.PAUSED_WHEN_CONNECTING) {
+      logger.debug("No reconnect - connector stopped, paused or paused_when_connecting");
+      return;
+    }
+
+    logger.debug("Reconnecting with load balancer...");
+    status = SingleConnectorStatus.CONNECTING;
+    connectWithLoadBalancer();
+    sendReport();
+  }
+
+  /**
+   * Prepares and sends heartbeat, or an initial report with information about the server. In the
+   * initial message includes server address, server port, list of services provided by the server
+   * and its weight, if set.
+   */
+  private void sendReport() {
+    ServerReport report;
+
+    if (status == SingleConnectorStatus.CONNECTING) { // prepare initial report
+      Server.Builder serverDetails =
+          com.rtbhouse.grpc.loadbalancer.Server.newBuilder()
+              .setIpAddress(ByteString.copyFrom(serverAddress.getAddress()))
+              .setPort(serverPort)
+              .addAllServices(Arrays.asList(services));
+
+      if (serverWeight != WEIGHT_NOT_DEFINED) {
+        serverDetails.setWeight(serverWeight);
+      }
+
+      report =
+          ServerReport.newBuilder()
+              .setReadyToServe(true)
+              .setServerDetails(serverDetails.build())
+              .build();
+
+      logger.info(
+          "Sending initial report: ready to serve true, server address {}, services {}, port {}, weight {}",
+          serverAddress.toString(),
+          String.join(", ", report.getServerDetails().getServicesList()),
+          serverPort,
+          report.getServerDetails().getWeight());
+
+    } else if (status == SingleConnectorStatus.WORKING) { // prepare heartbeat
+      report = ServerReport.newBuilder().setReadyToServe(true).build();
+      logger.debug("Sending heartbeat: ready to serve {}", report.getReadyToServe());
+
+    } else {
+      logger.warn(
+          "Requested to send report when connector status is neither CONNECTING nor WORKING "
+              + "(it is: {})",
+          status);
+      return;
+    }
+
+    try {
+      reportObserver.onNext(report);
+    } catch (IllegalStateException e) {
+      logger.info("Trying to send server report but the rpc has already ended");
+    }
+  }
+
+  /**
+   * Cleans up the current rpc that's being closed. Cancels sign-up if needed and calls
+   * onCompleted().
+   */
+  private synchronized void endRPC() {
+    logger.debug("Ending RPC...");
+
+    /* PAUSED_WHEN_CONNECTING state indicates, that connector is paused, but RPC is still opened,
+     * now we are ending RPC so we change state to PAUSED.
+     */
+    if (status == SingleConnectorStatus.PAUSED_WHEN_CONNECTING)
+      status = SingleConnectorStatus.PAUSED;
+
+    if (signUpHandle != null && !signUpHandle.isCancelled()) {
+      signUpHandle.cancel(true);
+    }
+    reportObserver.onCompleted();
+  }
+
+  /** Marks that the connector was paused. If needed, ends current rpc. */
+  synchronized void pause() {
```

Comment:
Ta pauza chyba nic nie robi sprytnego. Wysyłaczka heartbeatów jak rozumiem dalej próbuje wysyłać raporty, ale status jej na to nie pozwala. Może tutaj dałoby się wykorzystać tą flagę server_ready i ustawiać ją na false? Tzn. load balancer by widział serwerek, heartbeaty byłyby podtrzymywane, serwer nie brał by udziału w load balancingu do czasu zmiany tej flagi.

Reply:
Nie, pauza woła endRPC, więc woła się onCompleted co powoduje że LB natychmiast usuwa nas z listy serwerów, a wysyłaczka heartbeatów jest ubijana .cancel(). Można byłoby wykorzystywać tę flagę, ale chyba nie widać jakiejś znaczącej przewagi nad tym co jest teraz, więc chyba nie ma sensu tego ruszać.